rotkehlchen/exchanges/bitstamp.py

```diff
@@ -306,13 +301,17 @@ def _api_query(
             digestmod=hashlib.sha256,
         ).hexdigest()
 
-        self.session.headers.update({
+        headers = {
+            'X-Auth': f'BITSTAMP {self.api_key}',
+            'X-Auth-Version': 'v2',
             'X-Auth-Signature': signature,
             'X-Auth-Nonce': nonce,
             'X-Auth-Timestamp': timestamp,
-        })
+        }
         if content_type:
-            self.session.headers.update({'Content-Type': content_type})
+            headers.update({'Content-Type': content_type})
+
+        self.session.headers = CaseInsensitiveDict(headers)
```

Comment:
What is the significance of using a case insensitive dict here?

So the fix is essentially removing the `session` from `headers.update({'Content-Type': content_type})` right?

Because before the code was leaving it for other sessions too?

Reply:
The fix before I moved all the `headers` dict creation inside `_api_query()` was indeed either copying the existing dict or remove the content type; very similar to what we do with options. You are right asking below why remove `self.session.headers` if `X-Auth` and `X-Auth-Version` have always the same value.
However, after addressing bugs related with polluted dicts (i.e. options and headers) in different exchanges, I've thought that having these dicts data split in different places make us to overlook their updates and introduces bugs.

If you think it is overkilling I don't mind go back to updates and just take care of copying headers at the beginning of `_api_query()`.

Reply:
> If you think it is overkilling I don't mind go back to updates and just take care of copying headers at the beginning of _api_query().

Well assigning a key in every query without needing it is bad. So yes please remove the assignment of

1. `X-Auth-Version`. You can yourself see that you are assinging it to the constant `v2`. So why do it in each call? 
2. `X-Auth` this too is always assigned to the value of the API key, which is constant for the lifetime of the object. The API key is assigned only once at the constructor time.

For the rest, (signature, nonce and timestamp) you correctly re-assign them in every query! That's good.

---

~~But again you did not answer me ... what is the significant of CaseInsensitiveDict here? So that if the same header with different case somehow appears we don't consider it? I don't mind it ... just trying to understand why. ~~

Oh I see -- CaseInsensitiveHeaders is indeed for this, and it's the type used by Requests.